<h1>Introduction

A business owner of successful Indian chain restaurants wants to open his new branch in NY and NJ areas. 
He is not aware of the area and he wants to identify the best city in these two states to open the restaurants by using the density of already existing Indian restaurants in the area.

<h1>Data Section

For this project I will be using Four Square API.
I will be scouting for the Indian restaurants in New York City, Newark, Edison, Elizabeth, Secaucus. These are some of the cities in that geographical area with highest number of Indian communities and Indian restaurants.

<h1>Project Plan

For this project I will be using data from Foursquare API to assess the density of Indian restaurants through venues in the cities mentioned in Data Section.
From Four Square developer website I got the category of Indian Restaurants - 4bf58dd8d48988d10f941735.
I will be using the coordinates of these restaurants to map them on the folium maps imported for visual inspection.
Next, I will also try to find the centroid of these restaurants in every city.
Once I have the centroid Coordinates for each city, I will be calculating the distance of each restaurant from the centroid.

In [6]:
import numpy as np
import pandas as pd
import requests
from pandas.io.json import json_normalize 
print('All Libraries imported.')

All Libraries imported.


In [5]:
!conda install -c conda-forge folium=0.5.0 --yes
import folium
print('Folium installed and imported!')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    branca-0.4.1               |             py_0          26 KB  conda-forge
    ca-certificates-2020.6.20  |       hecda079_0         145 KB  conda-forge
    certifi-2020.6.20          |   py36h9f0ad1d_0         151 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    openssl-1.1.1g             |       h516909a_1         2.1 MB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    altair-4.1.0               |             py_1         614 KB  conda-forge
    ------------------------------------------------------------
                       

In [127]:
CLIENT_ID = 
CLIENT_SECRET = 
VERSION = '20180604'
LIMIT = 30
print('Success')

SyntaxError: invalid syntax (<ipython-input-127-bc48cc1102f1>, line 1)

In [73]:
LIMIT = 500
cities = ['New York, NY', 'Newark, NJ', 'Edison, NJ', 'Elizabeth, NJ', 'Secaucus, NJ']
results = {}
for city in cities:
    url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&near={}&limit={}&categoryId={}'.format(
        CLIENT_ID, 
        CLIENT_SECRET, 
        VERSION, 
        city,
        LIMIT,
        "4bf58dd8d48988d10f941735")
    results[city] = requests.get(url).json()

In [74]:
results

{'New York, NY': {'meta': {'code': 200,
   'requestId': '5f34495829b3b4547496ee41'},
  'response': {'suggestedFilters': {'header': 'Tap to show:',
    'filters': [{'name': '$-$$$$', 'key': 'price'},
     {'name': 'Open now', 'key': 'openNow'}]},
   'geocode': {'what': '',
    'where': 'new york ny',
    'center': {'lat': 40.742185, 'lng': -73.992602},
    'displayString': 'New York, NY, United States',
    'cc': 'US',
    'geometry': {'bounds': {'ne': {'lat': 40.882214, 'lng': -73.907},
      'sw': {'lat': 40.679548, 'lng': -74.047285}}},
    'slug': 'new-york-city-new-york',
    'longId': '72057594043056517'},
   'headerLocation': 'New York',
   'headerFullLocation': 'New York',
   'headerLocationGranularity': 'city',
   'query': 'indian',
   'totalResults': 254,
   'suggestedBounds': {'ne': {'lat': 40.82082780626791,
     'lng': -73.88633688247887},
    'sw': {'lat': 40.68321106215607, 'lng': -74.01899946794344}},
   'groups': [{'type': 'Recommended Places',
     'name': 'recommended

In [75]:
df_venues={}
for city in cities:
    venues = json_normalize(results[city]['response']['groups'][0]['items'])
    df_venues[city] = venues[['venue.id','venue.name', 'venue.location.address','venue.location.city','venue.location.state', 'venue.location.lat', 'venue.location.lng']]
    df_venues[city].columns = ['Id','Name', 'Address','City','State', 'Latt', 'Long']

In [72]:
df_venues

{'New York, NY':                           Id                                   Name  \
 0   4bbb9dbded7776b0e1ad3e51                       Tamarind TriBeCa   
 1   4a70a75bf964a52016d81fe3                    Bhatti Indian Grill   
 2   5272ca4511d22488f6895caf                         Drunken Munkey   
 3   510431dde4b0f2b954d00297                            Malai Marke   
 4   4593ed04f964a52050401fe3                  The Kati Roll Company   
 5   54f9f0e6498e4e163ca8937a        Deep Indian Kitchen (IndiKitch)   
 6   4b4aab62f964a520978c26e3                      Dil-e Punjab Deli   
 7   49d91c12f964a520015e1fe3                  The Kati Roll Company   
 8   43a7557bf964a5206b2c1fe3                 Punjabi Grocery & Deli   
 9   4d992c1e942ba093ebb5968c                              THELEwala   
 10  523b2b42498e1dfabcc8ab15                                Pippali   
 11  526fe5de498e115999f8e9cc                           Bengal Tiger   
 12  4a5b660df964a52015bb1fe3                   

In [65]:
print(df_venues)

{'New York, NY':                           Id                                   Name  \
0   4bbb9dbded7776b0e1ad3e51                       Tamarind TriBeCa   
1   4a70a75bf964a52016d81fe3                    Bhatti Indian Grill   
2   5272ca4511d22488f6895caf                         Drunken Munkey   
3   510431dde4b0f2b954d00297                            Malai Marke   
4   4593ed04f964a52050401fe3                  The Kati Roll Company   
5   54f9f0e6498e4e163ca8937a        Deep Indian Kitchen (IndiKitch)   
6   4b4aab62f964a520978c26e3                      Dil-e Punjab Deli   
7   49d91c12f964a520015e1fe3                  The Kati Roll Company   
8   43a7557bf964a5206b2c1fe3                 Punjabi Grocery & Deli   
9   4d992c1e942ba093ebb5968c                              THELEwala   
10  523b2b42498e1dfabcc8ab15                                Pippali   
11  526fe5de498e115999f8e9cc                           Bengal Tiger   
12  4a5b660df964a52015bb1fe3                                

In [120]:
maps = {}
for city in cities:
    city_latt = np.mean([results[city]['response']['geocode']['geometry']['bounds']['ne']['lat'],
                        results[city]['response']['geocode']['geometry']['bounds']['sw']['lat']])
    city_long = np.mean([results[city]['response']['geocode']['geometry']['bounds']['ne']['lng'],
                        results[city]['response']['geocode']['geometry']['bounds']['sw']['lng']])
    maps[city] = folium.Map(location=[city_lat, city_lng], zoom_start=12)

    for lat, lng, label in zip(df_venues[city]['Latt'], df_venues[city]['Long'], df_venues[city]['Name']):
        label = folium.Popup(label, parse_html=True)
        folium.CircleMarker(
            [lat, lng],
            radius=10,
            popup=label,
            color='blue',
            fill=True,
            fill_color='purple',
            fill_opacity=0.8,
            parse_html=False).add_to(maps[city])  
    print(f"Indian Resturants in {city} = ", results[city]['response']['totalResults'])

Indian Resturants in New York, NY =  254
Indian Resturants in Newark, NJ =  13
Indian Resturants in Edison, NJ =  60
Indian Resturants in Elizabeth, NJ =  37
Indian Resturants in Secaucus, NJ =  222


In [121]:
maps[cities[0]]

In [126]:
maps[cities[1]]

In [99]:
maps[cities[2]]

In [100]:
maps[cities[3]]

In [101]:
maps[cities[4]]

<b>Observation

Upon Observing Secacus NJ and NY City NY Maps , Its is clear that the data is overlapped and most of the resturants are in both sets of data. So its best to drop Secacus to drop from the Analysis

Now Lets find how close these resturants are in every city. By calculating the distance between Mean  coordinates and distance to every resturant in each city we can calcuate the Density.

In [124]:
for city in cities:
    city_lat = np.mean([results[city]['response']['geocode']['geometry']['bounds']['ne']['lat'],
                        results[city]['response']['geocode']['geometry']['bounds']['sw']['lat']])
    city_lng = np.mean([results[city]['response']['geocode']['geometry']['bounds']['ne']['lng'],
                        results[city]['response']['geocode']['geometry']['bounds']['sw']['lng']])
    maps[city] = folium.Map(location=[city_lat, city_lng], zoom_start=11)
    venues_mean_coor = [df_venues[city]['Latt'].mean(), df_venues[city]['Long'].mean()] 
    # add markers to map
    for Latt,Long, label in zip(df_venues[city]['Latt'], df_venues[city]['Long'], df_venues[city]['Name']):
        label = folium.Popup(label, parse_html=True)
        folium.CircleMarker(
            [Latt, Long],
            radius=5,
            popup=label,
            color='blue',
            fill=True,
            fill_color='#3186cc',
            fill_opacity=0.7,
            parse_html=False).add_to(maps[city])
        folium.PolyLine([venues_mean_coor, [Latt, Long]], color="green", weight=1.5, opacity=0.5).add_to(maps[city])
    
    label = folium.Popup("Mean Co-ordinate", parse_html=True)
    folium.CircleMarker(
        venues_mean_coor,
        radius=10,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(maps[city])

    print(city)
    print("Mean Distance from Mean coordinates")
    print(np.mean(np.apply_along_axis(lambda x: np.linalg.norm(x - venues_mean_coor),1,df_venues[city][['Latt','Long']].values)))

New York, NY
Mean Distance from Mean coordinates
0.025204772613466226
Newark, NJ
Mean Distance from Mean coordinates
0.015400081373466852
Edison, NJ
Mean Distance from Mean coordinates
0.026978970644616223
Elizabeth, NJ
Mean Distance from Mean coordinates
0.09812715045889935
Secaucus, NJ
Mean Distance from Mean coordinates
0.042743460153164846


In [125]:
maps[cities[0]]

In [108]:
maps[cities[1]]

In [109]:
maps[cities[2]]

In [110]:
maps[cities[3]]

In [111]:
maps[cities[4]]

As we have decided to Drop Secacuse from the Analysis we will Ignore the Map of City 4.

<h1>Conclusion

Indian Resturants in New York, NY =  254 \
Indian Resturants in Newark, NJ =  13 \
Indian Resturants in Edison, NJ =  60 \
Indian Resturants in Elizabeth, NJ =  37 \
Indian Resturants in Secaucus, NJ =  222 

New York, NY Mean Distance from Mean coordinates   0.025204772613466226\
Newark,   NJ Mean Distance from Mean coordinates   0.015400081373466852\
Edison,   NJ Mean Distance from Mean coordinates   0.026978970644616223\
Elizabeth,NJ Mean Distance from Mean coordinates   0.09812715045889935\
Secaucus, NJ Mean Distance from Mean coordinates   0.042743460153164846

If We compare the total number of resturants , New York stands on the top. But by the Density Newark has highly dense resturants. But the total number of resturants in Newark is 13. So its safe to ignore Newark. \
The Next Dense Ranking and Number of resturants go to Edison. \
 \
As New York is already crowded, its better if the Resturant is opened in Edison NJ


*By adding additional data points like - Asian Indian Population Density , Ratings of the existing Resturants etc this project's scope can be further expanded.*